src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+
+            if (taker_amount != place_offer)
+                crossed = true;
+
+            if (m_journal.debug)
+            {
+                m_journal.debug << "Offer Crossing: " << transToken (terResult);
+
+                if (terResult == tesSUCCESS)
+                {
+                    m_journal.debug <<
+                        "    takerPays: " << saTakerPays.getFullText () <<
+                        " -> " << place_offer.out.getFullText ();
+                    m_journal.debug <<
+                        "    takerGets: " << saTakerGets.getFullText () <<
+                        " -> " << place_offer.in.getFullText ();
+                }
+            }
+
+            saTakerPays = place_offer.out;
+            saTakerGets = place_offer.in;
+        }
+    }
+
+    if (terResult != tesSUCCESS)
+    {
+        m_journal.debug <<
+            "final terResult=" << transToken (terResult);
+
+        return terResult;
+    }
+
+    if (m_journal.debug)
+    {
+        m_journal.debug <<
+            "takeOffers: saTakerPays=" <<saTakerPays.getFullText ();
+        m_journal.debug <<
+            "takeOffers: saTakerGets=" << saTakerGets.getFullText ();
+        m_journal.debug <<
+            "takeOffers: mTxnAccountID=" <<
+            to_string (mTxnAccountID);
+        m_journal.debug <<
+            "takeOffers:         FUNDS=" << view.accountFunds (
+            mTxnAccountID, saTakerGets, fhZERO_IF_FROZEN).getFullText ();
+    }
+
+    if (saTakerPays < zero || saTakerGets < zero)
+    {
+        // Earlier, we verified that the amounts, as specified in the offer,
+        // were not negative. That they are now suggests that something went
+        // very wrong with offer crossing.
+        m_journal.fatal << (crossed ? "Partially consumed" : "Full") <<
+            " offer has negative component:" <<
+            " pays=" << saTakerPays.getFullText () <<
+            " gets=" << saTakerGets.getFullText ();
+
+        assert (saTakerPays >= zero);
+        assert (saTakerGets >= zero);
+        return tefINTERNAL;
+    }
+
+    if (bFillOrKill && (saTakerPays != zero || saTakerGets != zero))
+    {
+        // Fill or kill and have leftovers.
+        view.swapWith (view_checkpoint); // Restore with just fees paid.
+        return tesSUCCESS;
+    }
+
+    // What the reserve would be if this offer was placed.
+    auto const accountReserve (mEngine->getLedger ()->getReserve (
+        sleCreator->getFieldU32 (sfOwnerCount) + 1));
+
+    if (saTakerPays == zero ||                // Wants nothing more.
+        saTakerGets == zero ||                // Offering nothing more.
+        bImmediateOrCancel)                   // Do not persist.
+    {
+        // Complete as is.
+    }
+    else if (mPriorBalance.getNValue () < accountReserve)
+    {
+        // If we are here, the signing account had an insufficient reserve
+        // *prior* to our processing. We use the prior balance to simplify
```

Comment:
"simplify client writing" is ambiguous, and the natural English reading means "simplify writing that occurs in client" - whereas I think you mean "Make writing clients easier".

I'm not quite getting, actually, how this makes writing clients easier or simplifies the user experience?